pkg/rbd/controllerserver.go

```diff
@@ -265,93 +223,75 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
-		return nil, status.Error(codes.Internal, err.Error())
-	}
-
-	if err := cs.MetadataStore.Delete(volumeID); err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
-	}
-
-	delete(rbdVolumes, volumeID)
-	return &csi.DeleteVolumeResponse{}, nil
-}
-
-// ListVolumes returns a list of volumes stored in memory
-func (cs *ControllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
-	var startToken int
-	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_LIST_VOLUMES); err != nil {
-		klog.Warningf("invalid list volume req: %v", req)
-		return nil, err
-	}
-
-	//validate starting token if present
-	if len(req.GetStartingToken()) > 0 {
-		i, parseErr := strconv.ParseUint(req.StartingToken, 10, 32)
-		if parseErr != nil {
-			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
+		if _, ok := err.(ErrImageNotFound); !ok {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-		//check starting Token is greater than list of rbd volumes
-		if len(rbdVolumes) < int(i) {
-			return nil, status.Errorf(codes.Aborted, "invalid starting token %s", parseErr.Error())
-		}
-		startToken = int(i)
-	}
 
-	var entries []*csi.ListVolumesResponse_Entry
+		// If image is missing, then there was a key to point to it, it means we need to cleanup
+		// the keys and return success
+		volumeNameMutex.LockKey(rbdVol.RequestName)
+		defer func() {
+			if err := volumeNameMutex.UnlockKey(rbdVol.RequestName); err != nil {
+				klog.Warningf("failed to unlock mutex volume:%s %v", rbdVol.RequestName, err)
+			}
+		}()
 
-	keys := make([]string, 0)
-	for k := range rbdVolumes {
-		keys = append(keys, k)
+		if err := unreserveVol(rbdVol, req.GetSecrets()); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.DeleteVolumeResponse{}, nil
 	}
-	sort.Strings(keys)
 
-	for index, k := range keys {
-		if index < startToken {
-			continue
+	// lock out parallel create requests against the same volume name as we
+	// cleanup the image and associated omaps for the same
+	volumeNameMutex.LockKey(rbdVol.RequestName)
+	defer func() {
+		if err := volumeNameMutex.UnlockKey(rbdVol.RequestName); err != nil {
+			klog.Warningf("failed to unlock mutex volume:%s %v", rbdVol.RequestName, err)
 		}
-		entries = append(entries, &csi.ListVolumesResponse_Entry{
-			Volume: &csi.Volume{
-				VolumeId:      rbdVolumes[k].VolID,
-				CapacityBytes: rbdVolumes[k].VolSize,
-				VolumeContext: extractStoredVolOpt(rbdVolumes[k]),
-			},
-		})
-	}
+	}()
 
-	resp := &csi.ListVolumesResponse{
-		Entries: entries,
+	// Deleting rbd image
+	klog.V(4).Infof("deleting image %s", rbdVol.RbdImageName)
+	if err := deleteImage(rbdVol, rbdVol.AdminID, req.GetSecrets()); err != nil {
+		// TODO: can we detect "already deleted" situations here and proceed?
```

Comment:
If we dont see the image , thats a success.

Reply:
This is an [existing](https://github.com/ceph/ceph-csi/blob/f60a07ae8268136f40de3d62537f4f8b020f0ec4/pkg/rbd/controllerserver.go#L290) TODO, I am not touching this TODO yet, and do not intend to do so. Do you want me to fix this as a part of this patch?

Reply:
Removed the TODO, already addressed in code above, hence older TODO is no longer valid.